Rewrite entity collisions module.

JBYoshi opened the pull request:
Currently, the entity collisions module limits the return values of `World.getEntitiesWithinAABB()` and related methods to no more than 8 entries. However, this breaks many assumptions in both vanilla and mods - see #2452, SpongePowered/SpongeForge#1000, and SpongePowered/SpongeForge#1214.

This PR strips down the entity collisions module to do just what it was originally written for - entity collisions.

Fixes #2452.

JBYoshi committed:
Remove limits for getEntitiesWithinAABB for large bounding boxes. Fixes #2452

dualspiral commented:
I'm wondering if this is the right way to do it. What is the performance impact of continuously calculating AABB dimensions and entity size dimensions?

Does pigman angering happen in its own phase on the phase tracker? If so, would we be better off just using that phase to set `ignoreCollisions` to `true` and bypass this altogether? It feels less hacky then saying "the size of the AABB is larger than the largest entity".

JBYoshi commented:
> What is the performance impact of continuously calculating AABB dimensions and entity size dimensions?

I added some profiling code, and the dimension calculation averages about 40-60 nanoseconds for each call to `isLimitExempt()` on my computer. It can peak to about 200 ns or so during the first calls, but drops quickly.

> Does pigman angering happen in its own phase on the phase tracker?

Unfortunately, no. I believe it's part of `AttackEntityPacketState`, but that also uses one of the `getEntitiesWithinAABB` methods for some attacks.

dualspiral commented:
> I believe it's part of `AttackEntityPacketState`

In that case, should that ignore collisions instead? I'm trying to get a sense of where checking for this entity collisions optimisation makes sense - does it make sense during _that_ state?

JBYoshi commented:
It's more than just attacks. Some other cases that cause problems:
- Selectors. Distance-based selectors use `getEntitiesWithinAABB` for optimization.
- Structure blocks. This will prevent certain entities from being saved.
- Blocks that change size or position (such as farmland turning back to dirt, or pistons).

I'm wondering if it would be better to implement a whitelist of places where this behavior should take effect. The original purpose of the optimization was to limit [entity collisions](https://github.com/SpongePowered/SpongeCommon/commit/fbdedf182772129f1d1d147221c30320d607fe42). So maybe we could just replace the code for entity collisions; I think that would deal with the biggest of the performance drops.

In addition, many calls to `getEntitiesWithinAABB` just look for the closest entity that matches a predefined filter. If there was a way we could optimize for this case, it might be possible to rewrite those calls.

dualspiral commented:
> The original purpose of the optimization was to limit entity collisions. So maybe we could just replace the code for entity collisions; I think that would deal with the biggest of the performance drops.

We could. 

Is the code/process that we think the original intention is for in its own `PhaseState`? If so, perhaps:

* Default `ignoreCollisions` to `true`
* Set it to `false` for that specific phase state
* See what happens

If it is not, then maybe we could do that by intoducing a phase state. 

I think it would be worth keeping the optimisation code as arms-length as possible from the original target (as I think it is now) rather than rewriting specific calls for specific reasons, because there may be multiple areas that _could_ benefit from it.

gabizou commented on src/main/java/org/spongepowered/common/mixin/entitycollisions/ChunkMixin_Collisions.java:
```diff
@@ -124,4 +128,18 @@
+     * @author JBYoshi
+     */
+    private boolean collisionsImpl$isLimitExempt(AxisAlignedBB aabb) {
+        if (!(this.world instanceof WorldServer)) {
```
If we are keeping this logic, I'd much rather this be checking that `((WorldBridge) this.world).bridge$isFake()`

JBYoshi commented:
The more I look at this, the more I think the collisions module needs a complete rewrite. It looks like the original code already replaces most of the `collideWithNearbyEntities` behavior, to the point that the collision code may not even be running on an entity that's colliding with more than 8 other entities.

It might take some time, but I'll start playing around with the code and see what I can come up with.

Pulverizer commented:
This limit is also an issue for plugins as I attempted to use it and was wondering why my plugin wasn't detecting all the target entities. There was nothing in the JavaDocs to warn me that there is a limit in the config.

I have since switched to using `getNearbyEntities()` but I have no idea if this is actually worse for performance.

JBYoshi committed:
WIP rewrite of the collisions module

JBYoshi commented:
I've started on the rewrite for this. It fixes support for `/gamerule maxEntityCramming`, which should help by clearing out tightly-packed groups of entities. It also may fix some cases where `CollideEntityEvent` could be called for all AABB searches, although I haven't done any testing with that yet.

gabizou commented on src/main/java/org/spongepowered/common/mixin/core/entity/EntityLivingBaseMixin.java:
```diff
@@ -1170,4 +1171,12 @@ private void onResetActiveHand(final CallbackInfo ci) {
+    @Redirect(method = "onLivingUpdate", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/EntityLivingBase;collideWithNearbyEntities()V"))
+    private void collisions$runCollisions(EntityLivingBase self) {
+        try (PhaseContext<?> ignored = EntityPhase.State.COLLISION.createPhaseContext()
+                .source(this).buildAndSwitch()) {
```
Almost guaranteed that this might still be called on the client thread.

gabizou commented on src/main/java/org/spongepowered/common/mixin/core/entity/EntityLivingBaseMixin.java:
```diff
@@ -179,6 +179,7 @@
     @Shadow private boolean checkTotemDeathProtection(final DamageSource p_190628_1_) {
         return false; // SHADOWED
     }
+    @Shadow protected abstract void collideWithNearbyEntities();
```
prefix with `shadow$`.

gabizou commented on src/main/java/org/spongepowered/common/mixin/entitycollisions/BlockMixin_Collisions.java:
```diff
@@ -37,6 +37,9 @@
 
+// TODO this is currently inactive - No phase states for blocks override
+// isCollision, and CollideBlockEvent is handled with entities iterating
+// through blocks, not blocks iterating through entities
```
Then remove it.

gabizou commented on src/main/java/org/spongepowered/common/mixin/entitycollisions/ChunkMixin_Collisions.java:
```diff
@@ -99,27 +95,26 @@
                 return true;
             }
 
+            CollisionsCapability capability;
```
Should be `final`.

JBYoshi committed:
More updates.

- Removed block collision code.
- Removed default exceptions - I've tested all of them, and they work out of the box (see SpongePowered/SpongeForge#1000 and SpongePowered/SpongeForge#1214).
- Other miscellaneous fixes from gabizou's comments.

JBYoshi commented:
I've finished most of the cleanup, so this PR should be ready for a full review.

ImMorpheus reviewed (approved):
Working as intended

gabizou commented on src/main/java/org/spongepowered/common/mixin/core/entity/EntityLivingBaseMixin.java:
```diff
@@ -1170,4 +1171,16 @@ private void onResetActiveHand(final CallbackInfo ci) {
     // End implementation of UseItemStackEvent
 
+    @Redirect(method = "onLivingUpdate", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/EntityLivingBase;collideWithNearbyEntities()V"))
+    private void collisions$runCollisions(EntityLivingBase self) {
```
If this is supposed to be added when the collision plugin has enabled collisions, this should be in a separate mixin that gets applied.

ImMorpheus replied:
This is necessary to enter the `EntityCollisionState` and it should always be applied. 
The name was probably copied from the old Mixin.

JBYoshi replied:
I've renamed it to `impl$runCollisions`.

JBYoshi committed:
Rename collisions to impl$.

JBYoshi committed:
Merge remote-tracking branch 'upstream/stable-7' into entity-collision-large-boxes
